GitUI/CommandsDialogs/FormBrowse.cs

```diff
@@ -1795,26 +1795,47 @@ private void CurrentBranchDropDownOpening(object sender, EventArgs e)
         {
             branchSelect.DropDownItems.Clear();
 
-            ToolStripMenuItem item = new ToolStripMenuItem(checkoutBranchToolStripMenuItem.Text);
-            item.ShortcutKeys = checkoutBranchToolStripMenuItem.ShortcutKeys;
-            item.ShortcutKeyDisplayString = checkoutBranchToolStripMenuItem.ShortcutKeyDisplayString;
-            branchSelect.DropDownItems.Add(item);
-            item.Click += (hs, he) => CheckoutBranchToolStripMenuItemClick(hs, he);
-
+            AddCheckoutBranchMenuItem();
             branchSelect.DropDownItems.Add(new ToolStripSeparator());
+            AddBranchesMenuItems();
 
-            foreach (var branch in Module.GetRefs(false))
+            PreventToolStripSplitButtonClosing(sender as ToolStripSplitButton);
+        }
+        private void AddCheckoutBranchMenuItem()
+        {
+            var checkoutBranchItem = new ToolStripMenuItem(checkoutBranchToolStripMenuItem.Text)
+            {
+                ShortcutKeys = checkoutBranchToolStripMenuItem.ShortcutKeys,
+                ShortcutKeyDisplayString = checkoutBranchToolStripMenuItem.ShortcutKeyDisplayString
+            };
+            branchSelect.DropDownItems.Add(checkoutBranchItem);
+            checkoutBranchItem.Click += CheckoutBranchToolStripMenuItemClick;
```

Comment:
I suggest to unwire all events in `DropDownClosed` otherwise everytime a user clicks to checkout a branch (or just to see what branches are available) we are creating new objects and thus causing unnecessary GC runs.

Reply:
Can you explain this a bit more?

Currently the `DropDownOpening` clears all the previously existing drop-down items, and I assume that when this is done, since they are no longer referenced, they are collected by the GC, as well as the binded events. Is this correct? In that case, it's almost the same as if I unbind the event myself in the `DropDownClosed`, isn't it?

Thanks for your time explaining this.

Reply:
GC runs are non-deterministic, they do happen but you don't control when (unless you call it expressly). They are also reasonably expensive, since GC has to pause the execution and walk the ref tree, release memory, execute the finalizer queue, promote generations, compact etc...

So by not expressly unbinding events, and thus releasing resources associated with them, you delegate that work to GC. ...and you're causing a memory leak.
MSDN states that in https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events (scroll to "Unsubscribing" section). Jon Skeet seems to think otherwise (https://stackoverflow.com/a/17400033), but I tend to trust MS a little more.

Now imagine a user constantly opening and closing the dropdown - the will be tens upon tens of abandoned handlers.... I imagine this will be causing constant GC runs, and ultimately lead to degradation of the app's performance. A memory profile test is in order, but I don't have time for this right now.

I was taught many years ago - *if you open something, make sure to close it*. Granted back in the days we didn't have the luxury of GC.

Reply:
@RussKie hmm, in this case unsubscribing would add unnecessary code. The MS article you linked to lists reasons for unsubscribing such as "prevent double firing" and "prevent leaks when parent objects are disposed". These situations are not applicable here.
The delegate for this event will be GCed at the same time as the ToolStripMenuItem which holds it.

The second part of the reasoning in the article about object not being GCed while event handled has a reference to it is also not applicable here as in this concrete case it would mean "FormBrowse would not be GCed until ToolStripMenuItem has a reference to it" which is fine as FormBrowse lives longer than the menu items.

Reply:
I suppose you are right.
Would be interesting to profile this some time....

Reply:
Thanks for your comments and the links provided. I guess that it's not necessary to explicitly unsubscribe after all, so I will leave this code unchanged.